transaction-client/src/main/java/org/apache/omid/tso/client/TSOClient.java

```diff
@@ -513,7 +515,7 @@ void decrementRetries() {
         }
 
         public StateMachine.State handleEvent(CloseEvent e) {
-            factory.releaseExternalResources();
+            bootstrap.config().group().shutdownGracefully();
```

Comment:
If you have two racing `CloseEvent` messages, does `shutdownGracefully()` handle this correctly (i.e. is synchronized and exits quietly on the 2nd).